packages/main/src/plugin/kubernetes-context-state.ts

```diff
@@ -0,0 +1,223 @@
+  deploymentInformer?: Informer<V1Deployment> & ObjectCache<V1Deployment>;
+}
+
+// ContextState stores information for the user about a kube context: is the cluster reachable, the number
+// of instances of different resources
+export interface ContextState {
+  reachable: boolean;
+  podsCount: number;
+  deploymentsCount: number;
+}
+
+// the ContextsState singleton (instantiated by the kubernetes-client singleton)
+// manages the state of the different kube contexts
+export class ContextsState {
+  private kubeConfig = new KubeConfig();
+  private contextsState = new Map<string, ContextState>();
+  private contextsInternalState = new Map<string, ContextInternalState>();
+
+  constructor(private readonly apiSender: ApiSenderType) {}
+
+  // update is the reconcile function, it gets as input:
+  // - the user preference indicating if the user wants to get live information about kube contexts
+  // - the last known kube config
+  // and starts/stops informers for different kube contexts, depending on these inputs
+  async update(kubeconfig: KubeConfig) {
+    this.kubeConfig = kubeconfig;
+    // Add informers for new contexts
+    for (const context of this.kubeConfig.contexts) {
+      if (this.contextsInternalState.get(context.name) === undefined) {
+        this.contextsState.set(context.name, {
+          reachable: false,
+          podsCount: 0,
+          deploymentsCount: 0,
+        });
+        const informers = this.createKubeContextInformers(context);
+        if (informers) {
+          this.contextsInternalState.set(context.name, informers);
+        }
+      }
+    }
+    // Delete informers for removed contexts
+    let removed = false;
+    for (const [name, state] of this.contextsInternalState) {
+      if (!this.kubeConfig.contexts.find(c => c.name === name)) {
+        await state.podInformer?.stop();
+        await state.deploymentInformer?.stop();
+        this.contextsInternalState.delete(name);
+        this.contextsState.delete(name);
+        removed = true;
+      }
+    }
+    if (removed) {
+      this.dispatchContextsState();
+    }
+  }
+
+  createKubeContextInformers(context: KubeContext): ContextInternalState | undefined {
+    const kc = new KubeConfig();
+    const cluster = this.kubeConfig.clusters.find(c => c.name === context.cluster);
+    if (!cluster) {
+      return;
+    }
+    const user = this.kubeConfig.users.find(u => u.name === context.user);
+    if (!user) {
+      return;
+    }
+    kc.loadFromOptions({
+      clusters: [cluster],
+      users: [user],
+      contexts: [context],
+      currentContext: context.name,
+    });
+
+    const ns = context.namespace ?? 'default';
+    return {
+      podInformer: this.createPodInformer(kc, ns, context),
+      deploymentInformer: this.createDeploymentInformer(kc, ns, context),
+    };
+  }
+
+  private createPodInformer(kc: KubeConfig, ns: string, context: KubeContext): Informer<V1Pod> & ObjectCache<V1Pod> {
```

Comment:
There is a lot of duplication between createPodInformer and createDeploymentInformer. Could you refactor out a common function sort of like kubernetes-client.ts makeKubernetesInformer?

Reply:
There are many smaller duplications like:

```
      const previous = this.contextsState.get(context.name);
      if (previous) {
        /*
           do something with previous context state
        */
      }
      this.dispatchContextsState();
```